metadata-ingestion/tests/unit/test_transform_dataset.py

```diff
@@ -1822,6 +1825,209 @@ def fake_get_domain(entity_urn: str) -> models.DomainsClass:
     assert server_domain in transformed_aspect.domains
 
 
+def test_pattern_add_dataset_domain_match(mock_datahub_graph):
+    acryl_domain = builder.make_domain_urn("acryl.io")
+    gslab_domain = builder.make_domain_urn("gslab.io")
+    pattern = "urn:li:dataset:\\(urn:li:dataPlatform:bigquery,.*"
+
+    pipeline_context: PipelineContext = PipelineContext(
+        run_id="test_simple_add_dataset_domain"
+    )
+    pipeline_context.graph = mock_datahub_graph(DatahubClientConfig)
+
+    output = run_dataset_transformer_pipeline(
+        transformer_type=PatternAddDatasetDomain,
+        aspect=models.DomainsClass(domains=[gslab_domain]),
+        config={
+            "domain_pattern": {"rules": {pattern: [acryl_domain]}},
+        },
+        pipeline_context=pipeline_context,
+    )
+
+    assert len(output) == 2
+    assert output[0] is not None
+    assert output[0].record is not None
+    assert isinstance(output[0].record, MetadataChangeProposalWrapper)
+    assert output[0].record.aspect is not None
+    assert isinstance(output[0].record.aspect, models.DomainsClass)
+    transformed_aspect = cast(models.DomainsClass, output[0].record.aspect)
+    assert len(transformed_aspect.domains) == 2
+    assert gslab_domain in transformed_aspect.domains
```

Comment:
So PATCH is the default? (if semantics are not specified)

Reply:
No, OVERWRITE is the default, but if you don't have `replace_existing = True` it appends new domains to the existing domains for the entity. https://datahubproject.io/docs/metadata-ingestion/docs/transformer/dataset_transformer#relationship-between-replace_existing-and-semantics explains all the possible situations, although I don't like how this is kinda confusing

Reply:
Ah this is quite confusing... 

So IF the aspect is a DomainsClass aspect, then an OVERWRITE will mean "we merge into this aspect".

If the aspect is not a DomainsClass aspect, then we'll just generate a new DomainsClass aspect altogether. 

I think this makes sense but the documentation is a bit hard to follow for sure.

Maybe another set of cases that would be useful to include in the tests are what happens when a non-domains aspect (e.g. DatasetPropertiesClass) is sent through the pipeline (to show how it behaves then)

Reply:
The way I see it, the semantics just says whether you edit the existing row (via a `AddDatasetDomain` aspect) or create a new one (via a `DomainsClass` aspect). `replace_existing` determines whether you append to the existing domains list, or truncate. And, confusingly, pattern matching does absolutely nothing to change this logic -- transformers are applied on all change event / proposals with matching `aspect_name`, the pattern matching just allows you to say which entities get which domain, and if there's no match then they get no domains.

Shouldn't the pipeline around transformers (i.e. `base_transformer.py::BaseTransformer.transform` and `_transform_or_record_mcp` in that same class) handle that filtering? I would expect testing around that to be on the base transformer, rather than every transformer than inherits from the base transformer.

Reply:
Hmm my understanding of this transformer is that it can handle ANY aspects not just incoming domain aspects produced by the Ingestion Source... Maybe that's a bad interpretation

Reply:
It seems that way. So for example, if I configure a Transformer with an Ingestion Source that does not itself sent DomainsClass aspects (some do not), then this transformer will intercept whatever aspect it _does_ produce (e.g. datasetProperties) and will inject the Domains aspect (this would be my expectation, at least)

Reply:
Got it - so yeah the parent class adds filtering on the Domains aspect... Very odd because the source should not need to produce a Domains aspect for this to apply... Let me see if my understanding of the docs aligns